### PR TITLE
[Dashboard] Enable dashboard in the minimal ray installation

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This is the last PR to enable dashboard in the minimal ray installation.

Look https://docs.google.com/document/d/12qP3x5uaqZSKS-A_kK0ylPOp0E02_l-deAbmm8YtdFw/edit# for more details;

Screen Shot 2022-01-27 at 6 00 39 AM

TODO
- [ ] Add a minimal test to verify dashboard being started.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/dashboard.py [hunk 1]
```diff
@@ -235,26 +177,19 @@ async def run(self):
...
-        # Need daemon True to avoid dashboard hangs at exit.
-        service_discovery.daemon = True
-        service_discovery.start()
+            log_dir=args.log_dir,
+            temp_dir=args.temp_dir,
+            minimal=args.minimal)
         loop = asyncio.get_event_loop()
         loop.run_until_complete(dashboard.run())
     except Exception as e:
+        # logger.exception(e)
```
Comment (rkooo567) [inline dashboard/dashboard.py:186]: I will revert this back

File: dashboard/head.py [hunk 2]
```diff
@@ -302,3 +273,6 @@ async def _async_notify():
         await asyncio.gather(*concurrent_tasks,
                              *(m.run(self.server) for m in modules))
         await self.server.wait_for_termination()
+
+        if not self.minimal:
```
Comment (rkooo567) [inline dashboard/head.py:277]: I will change this to if self.http_server

File: dashboard/modules/reporter/reporter_head.py [hunk 3]
```diff
@@ -31,6 +32,20 @@ def __init__(self, dashboard_head):
         self._stubs = {}
         self._ray_config = None
         DataSource.agents.signal.append(self._update_stubs)
+        # TODO(fyrestone): Avoid using ray.state in dashboard, it's not
+        # asynchronous and will lead to low performance. ray disconnect()
+        # will be hang when the ray.state is connected and the GCS is exit.
+        # Please refer to: https://github.com/ray-project/ray/issues/16328
+        assert dashboard_head.gcs_address or dashboard_head.redis_address
```
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_head.py:39]: Reporter agent is the place to aggregate metrics, so I added the prometheus writing logic into reporter head

File: dashboard/head.py [hunk 4]
```diff
@@ -140,6 +135,21 @@ def __init__(self, http_host, http_port, http_port_retries, gcs_address,
             self.server, f"{grpc_ip}:0")
         logger.info("Dashboard head grpc address: %s:%s", grpc_ip,
                     self.grpc_port)
+        # If the dashboard is started as non-minimal version, http server should
+        # be configured to communicate with the dashboard in a head node.
+        self.http_server = None
```
[the remark this answers is not in the document]
Comment (jjyao) [reply dashboard/head.py:140]: Seems this comment is saying that http server should be configured to communicate with itself?

[the remark this answers is not in the document]
Comment (rkooo567) [reply dashboard/head.py:140]: Oh, I think I just copied this part from agent server and haven't fixed the comment... :(. Let me make a change here!

**Comment by rkooo567** [comment]
@jjyao Here's how the classes are structured! 

head node has a dashboard process
dashboard process starts from dashboard.py, which instantiates `DashboardHead`
Dashboard process contains *_head.py as modules. (so those code doesn't run inside agents).

**Comment by rkooo567** [comment]
I will address comments tmrw

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by rkooo567** [comment]
This is ready for review again.

**Comment by rkooo567** [comment]
cc @bveeramani do you know why lint is failing? The local lint & CI lint seems to have different formatting. 

Is there a specific Black version I should download?

**Review by edoakes — APPROVED** [review]

**Comment by rkooo567** [comment]
Failed tests are flaky in the master

**Pull request merged**